### DISCUSSION TITLE
Unable to read the content of `WebSocketUpgradeError.response`

### DISCUSSION THREAD

**Discussion by WSH032** [frankie567/httpx-ws#69] [Q&A]
## Describe the bug

Unable to read the content of `WebSocketUpgradeError.response`, because the stream has already been closed.

## To Reproduce

Launch a server to send a denial handshake response.

```py3
from fastapi import FastAPI, WebSocket
from fastapi.responses import Response
from uvicorn import run

app = FastAPI()


@app.websocket("/")
async def _(ws: WebSocket):
    resp = Response(content="foo")
    await ws.send_denial_response(resp)


run(app)
```

Next, attempt the handshake using `httpx-ws` and read the response content after the handshake fails.

```py3
from contextlib import AsyncExitStack

import anyio
from httpx_ws import WebSocketUpgradeError, aconnect_ws


async def main():
    async with AsyncExitStack() as exit_stack:
        try:
            await exit_stack.enter_async_context(aconnect_ws("ws://127.0.0.1:8000/"))
        except WebSocketUpgradeError as e:
            # httpx.StreamClosed: Attempted to read or stream content, but the stream has been closed.
            assert await e.response.aread() == b"foo"


anyio.run(main)
```

Then you can see following error:

```shell
  File "e:\GitHub\fastapi-proxy\Untitled-3.py", line 12, in main
    await e.response.aread()
  File "E:\GitHub\fastapi-proxy\.venv\lib\site-packages\httpx\_models.py", line 911, in aread
    self._content = b"".join([part async for part in self.aiter_bytes()])
  File "E:\GitHub\fastapi-proxy\.venv\lib\site-packages\httpx\_models.py", line 911, in <listcomp>
    self._content = b"".join([part async for part in self.aiter_bytes()])
  File "E:\GitHub\fastapi-proxy\.venv\lib\site-packages\httpx\_models.py", line 929, in aiter_bytes
    async for raw_bytes in self.aiter_raw():
  File "E:\GitHub\fastapi-proxy\.venv\lib\site-packages\httpx\_models.py", line 978, in aiter_raw
    raise StreamClosed()
httpx.StreamClosed: Attempted to read or stream content, but the stream has been closed.
```

## Expected behavior

The propagation of `WebSocketUpgradeError` leads to the exit of the `client.stream` context manager, then the `response` content is no longer readable.

https://github.com/frankie567/httpx-ws/blob/940c9adb3afee9dd7c8b95514bdf6444673e4820/httpx_ws/_api.py#L1048-L1050

Can we make a compromise: request the user to manually close the response when `WebSocketUpgradeError` is raised, instead of closing it automatically?

```py3
except WebSocketUpgradeError as e:
    await e.aclose()
```

## Configuration
 ```
fastapi == 0.110.1
uvicorn == 0.29.0
httpx-ws == 0.6.0
```

**Comment by frankie567** [comment]
Hmm, I see! Interesting case 😅

I would prefer to keep the normal functioning of the context manager, to keep a good developer experience.

Before raising the exception, could we actually read the content? I believe then [HTTPX will cache it](https://github.com/encode/httpx/blob/45bb65bba12360d94b8c512e6b13ac4b775a402d/httpx/_models.py#L908-L914) and make it available immediately on further reads. Something like:

```py
    async with client.stream("GET", url, headers=headers, **kwargs) as response:
        if response.status_code != 101:
            await response.aread()
            raise WebSocketUpgradeError(response)
```

What do you think?

**Reply by frankie567** [reply]
EDIT: I had a closer look at your stack trace and saw this:

```
  File "E:\GitHub\fastapi-proxy\.venv\lib\site-packages\httpx_ws\_api.py", line 1183, in _aconnect_ws
    print(response.read())
```

Looks like you added a manual print there. You call `response.read()`, while working with an async response (should be `await response.aread()` then). I believe this is the root cause of the error you see. Try to remove this and tell me if you still encounter the issue you describe.

**Reply by WSH032** [reply]
Looks like you added a manual print there. You call [response.read](https://response.read/)(), while working with an async response (should be await response.aread() then). I believe this is the root cause of the error you see. Try to remove this and tell me if you still encounter the issue you describe.

Oops, I copied the wrong error message, sorry.
The actual error is: httpx.StreamClosed: Attempted to read or stream content, but the stream has been closed.
I've updated the previous traceback.

Before raising the exception, could we actually read the content? I believe then [HTTPX will cache it](https://github.com/encode/httpx/blob/45bb65bba12360d94b8c512e6b13ac4b775a402d/httpx/_models.py#L908-L914) and make it available immediately on further reads. Something like:

Indeed, httpx does cache it, but this increases IO time, and not every user has the need to read the content.
Moreover, in the case of a huge response body, such as with file responses, it would occupy a huge amount of memory.

**Reply by frankie567** [reply]
> Indeed, httpx does cache it, but this increases IO time, and not every user has the need to read the content.

True but, in this context, it would be quite unexpected to have a big response body, so I would say that's acceptable 🤔 

<sup><sub>*Sorry, I edited your previous message by mistake; I thought I clicked on *Quote* ☺️</sub></sup>

**Reply by WSH032** [reply]
> in this context, it would be quite unexpected to have a big response body

But if the user use an incorrect URL, leading them to believe they are establishing a WebSocket request, while in reality, they are making an HTTP file request, this situation can occur.

for example:
```py
aconnect_ws("https://speed.hetzner.de/100MB.bin")
```

**Reply by WSH032** [reply]
I just discovered that `httpx-ws` doesn't seem to perform handshake checks, referring to the [wiki](https://en.wikipedia.org/wiki/WebSocket#Protocol_handshake):

> In addition to Upgrade headers, the client sends a Sec-WebSocket-Key header containing [base64](https://en.wikipedia.org/wiki/Base64)-encoded random bytes, and the server replies with a [hash](https://en.wikipedia.org/wiki/Hash_function) of the key in the Sec-WebSocket-Accept header. This is intended to prevent a [caching](https://en.wikipedia.org/wiki/Cache_(computing)) [proxy](https://en.wikipedia.org/wiki/HTTP_proxy) from re-sending a previous WebSocket conversation,[[37]](https://en.wikipedia.org/wiki/WebSocket#cite_note-38) and does not provide any authentication, privacy, or integrity. The hashing function appends the fixed string 258EAFA5-E914-47DA-95CA-C5AB0DC85B11 (a [UUID](https://en.wikipedia.org/wiki/UUID)) to the value from Sec-WebSocket-Key header (which is not decoded from base64), applies the [SHA-1](https://en.wikipedia.org/wiki/SHA-1) hashing function, and encodes the result using base64.[[38]](https://en.wikipedia.org/wiki/WebSocket#cite_note-39)

Currently, `httpx-ws` only checks for the 101 status code but does not verify the `Sec-WebSocket-Key`.

But this is another issue. Even if the handshake check is done correctly, if the server insists on sending a large response body, it can still bite the user.

**Reply by frankie567** [reply]
> Currently, httpx-ws only checks for the 101 status code but does not verify the Sec-WebSocket-Key.

Indeed 🙃 Feel free to open an issue about this.

> But this is another issue. Even if the handshake check is done correctly, if the server insists on sending a large response body, it can still bite the user.

True, true... 🤔 What we could do is to provide a lower-level API where you could have more control of the stream so you can read the response body. However, we would keep the current API like this for which, in 99% of the case, people won't even care about the response.

In a sense, it's quite similar to what HTTPX does for `stream`: https://www.python-httpx.org/async/#streaming-responses

**Reply by WSH032** [reply]
> in 99% of the case, people won't even care about the response.

I agree with you

---

> In a sense, it's quite similar to what HTTPX does for stream: https://www.python-httpx.org/async/#streaming-responses

Do you mean that?

> For situations when context block usage is not practical, it is possible to enter "manual mode" by sending a [Request instance](https://www.python-httpx.org/advanced/clients/#request-instances) using client.send(..., stream=True).
>
> Example in the context of forwarding the response to a streaming web endpoint with [Starlette](https://www.starlette.io/):

I'm doing something similar. I'm attempting to implement a transparent WebSocket proxy server and encountered the issue being discussed during testing.

https://github.com/WSH032/fastapi-proxy-lib/blob/61192fcdccf4e7c85f65db2394b06d145f6f9b25/src/fastapi_proxy_lib/core/websocket.py#L414-L441
